_posts/2021-04-15-testing-react-tracking-with-jest-and-enzyme.markdown

```diff
@@ -0,0 +1,261 @@
+date: 2021-04-15
+categories: [testing, tracking, volt, cms, enzyme, react-tracking]
+author: matt-dole
+comment_id: 680
+---
+
+Recently, I needed to test a button that would make an analytics tracking call using
+[react-tracking](https://github.com/NYTimes/react-tracking) and then navigate to a new page in a callback. This
+presented some challenges - I wasn't sure how to create a mocked version of react-tracking that would allow a
+callback to be passed.
+
+With some help from fellow Artsy engineers [Christopher Pappas](https://github.com/damassi) and
+[Pavlos Vinieratos](https://twitter.com/pvinis), I got the tracking and testing to work. Here's how we did it.
+
+<!-- more -->
+
+# A little context
+
+This work took place in Volt, our partner CMS (it's sadly a private repository, but I'll do my best to paste in
+relevant code snippets so you're not totally in the dark). Volt has been around for a long time and has had several
+different client-side tracking implementations over the years. In this case, I wanted to take the opportunity to
+bring Volt up to standard with our other big apps, [Force](https://github.com/artsy/force/) and
+[Eigen](https://github.com/artsy/eigen). They both use react-tracking and
+[Cohesion](https://github.com/artsy/cohesion/), our centralized analytics schema.
+
+Our use-case was a button that would navigate the user to a new page. The button had been implemented in a previous
+PR, and now we wanted to make it execute a tracking call before navigating.
+
+We use Segment for tracking, and their tracking setup relies on a JS snippet being available on your pages. That
+snippet sets a `window.analytics` property, which in turn
+[has a](https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#track) `.track()`
+method. On a fundamental level, all of our tracking calls boil down to a call to `window.analytics.track()`. We
+pass a list of properties to `.track()`, Segment receives the event and properties, and the resulting JSON is
+stored in our data warehouse.
+
+# Adding react-tracking
+
+First, there was a bit of setup required to get react-tracking working. The react-tracking package
+[assumes you're using Google Tag Manager by default](https://github.com/NYTimes/react-tracking#custom-optionsdispatch-for-tracking-data),
+but allows you to override that behavior with a custom `dispatch` function. In our case, we wrap our React apps in
+a `<BaseApp>` component, so we added a new `<TrackerContextCohesion>` component with a custom `dispatch` that would
+be available to all of our React apps:
+
+```ts
+import React from "react"
+import { track } from "react-tracking"
+
+const trackEvent = ({ data, options, callback }) => {
+  // Action can be something like "click" or "Viewed tooltip"; we pull it out and send it to Segment
+  // separately since it defines the event type
+  const actionName = data.action || data.action_type
+  const trackingData = omit(data, ["action_type", "action"])
```

Comment:
We can remove this stuff as its very internal artsy specific. And looking closely, this whole function isn't needed since its just a wrapper for artsy's weird legacy vs modern tracking schema migration.

Can we just put everything in the dispatch below? Something like: 

```tsx
dispatch: ({ data, options, callback }) => {
  window.analytics.track(data.action, data, options, callback)
},
```

Reply:
Good call! making that update now

Reply:
@mdole - were you going open a new PR for this?